Make FoV routine

aburrell opened the pull request:
A preliminary version of the routine that automatically determines the backscatter return direction.  This partially addresses #135.

Test results using a fitacf 3.0 file:

`make_fov -cn a ~/Programs/Data/SuperDARN/Downloads/19960916.0400.00.han.fitacf3 > han.tst_fov`

This creates an ascii file with 13727 lines and the following columns:

```
#STID DATE TIME BMNUM BMAZM CPID INTT_SC INTT_US NAVE FRANG RSEP RXRISE FREQ NOISE ATTEN CHANNEL NRANG RG GFLG FOVFLG FOV_PAST GRPFLG GRPNUM GRPID P_0 P_0_ERR V V_ERR W_L W_L_ERR P_L P_L_ERR PHI0 PHI0_ERR ELV ELV_LOW ELV_HIGH VH VH_ERR VH_METHOD REGION HOP DIST MED_P_0 MED_P_0_ERR MED_V MED_V_ERR MED_W_L MED_W_L_ERR MED_P_L MED_P_L_ERR MED_PHI0 MED_PHI0_ERR OPP_ELV OPP_ELV_LOW OPP_ELV_HIGH OPP_VH OPP_VH_ERR OPP_VH_METHOD OPP_REGION OPP_HOP OPP_DIST
```

Re-creating one of the figures from my paper leads to similar results.  This scan is from a different time that was created during various spot checks.

![fov_fig4_test_third_scan](https://user-images.githubusercontent.com/7045886/128564537-28b2745e-aba8-406b-8cfa-f7c1f928fe86.png)

Before moving this from draft status we need to:
- [x] merge #444 
- [x] Move general utility functions into appropriate libraries
- [x] Create a fitacf-esq non-ascii output file option
- [x] Finalize command-line options

aburrell committed:
commented subroutine

Commented the exclude_outofscan subroutine.

aburrell committed:
Update make_grid.c

Added more descriptions to the subroutine headers

aburrell committed:
Create read_fitacf.c

First draft of a routine to read fit binaries to stdout

aburrell committed:
file rename

Renamed file to better reflect functionality

aburrell committed:
Create grid_funcs.c

Copied the neccessary functions from make_grid.  Need these to be somewhere common in the future.

aburrell committed:
Create write_scan.c

Created routine to write out fit data in a human readable format.

aburrell committed:
Update read_fit.c

Updated file to not contain the duplicate grid routines and use the new write function

aburrell committed:
Create makefile

Created a makefile for the new procedure

aburrell committed:
renamed directory

Renamed directory to match binary

hannahhosley committed:
added median_fit_v1.0

Added the basis for a median fitting routine that uses fit data.

aburrell committed:
Removed unnecessary routine

Removed unnecessary routine created for summer student.

aburrell committed:
Update write_scan.c

Added writing routine that cycles through new structure.

aburrell committed:
Create multscan.h

Created a new structure that will be needed for FoV calculation.

aburrell committed:
updated routines

Added routines to support new fit structure

aburrell committed:
Update read_fit.c

Fixed bug in end of line.

Commit:
Fixed multscan.h file not found error

Moved multscan.h to the inlcude folder to follow the typical
rst library structure and then modified the makefile to allow it
to be compiled correctly.

aburrell committed:
LCL: update .gitignore to prevent messy commits

Updated the .gitignore to ignore the local user changes needed to install
RST.

aburrell committed:
LCL: update .gitignore

Removed lines that weren't working.

aburrell committed:
BUG: compilaton bugs

Fixed errors and warnings present during compilation.  Added comments to
help ensure proper local libararies are included in the future.

aburrell committed:
STY: trailing whitespace in makefile

Removed trailing whitespace from makefiles.

aburrell committed:
STY: radar.c

Added some whitespace to separate testing and reading parts of routine.

aburrell committed:
ENH: moved read_fit_v1.0 to tools

Moved read_fit from library to tools directory, a more appropriate location.

aburrell committed:
ENH: generalized loading routines

Generalized the loading routine, to either write the data line-by-line or store
the data in a multi-scan linked list.

aburrell committed:
ENH: removed unused libraries

Removed unused libraries.

aburrell committed:
STY: whitespace

Removed trailing whitespace in makefile

aburrell committed:
ENH: Added make_fov

Added the outline of the FoV routine.  To avoid adding more data structures
than necessary, included all of the data and flags that will be needed for the
backscatter ID code as well.

aburrell committed:
ENH: Added make/free routines

Added make/free routines for the multscan structures.

aburrell committed:
ENH: added elevation routine

Adapted Simon's elevation routine to incude a field of view flag.

aburrell committed:
ENH: added single-beam groundscatter eval

Added the functions needed to perform the single-beam groundscatter flag evaluation.

aburrell committed:
ENH: added autosave pattern to gitignore

Added another autosave pattern to gitignore.

aburrell committed:
BUG: debugged read_fit

Improved read_fit output.

aburrell committed:
ENH: updated structures

Refined the parameters needed for this data structure.

aburrell committed:
ENH: improved make_fov outline

Further fleshed out the make_fov routine outline.

aburrell committed:
ENH: created update_beam frame

Created a frame for the update_beam routine.

aburrell committed:
BUG: fixed DCE hardware file

Fixed the bug in the DCE hardware file for local processing.

aburrell committed:
ENH: started update_scan framework

Started the update_scan framework, needs an overhaul.

aburrell committed:
ENH: adjust multbsid struct

Adjust the multbsid structure to include better parameters.

aburrell committed:
DEL: remove duplicate code

Remove duplicated elevation angle code.

aburrell committed:
BUG: fix eval_groundscatter inputs

Fixed the inputs to the `EvalGroundScatter` routine.

aburrell committed:
ENH: added partial eval_scan_fov code

Added a partial implementation of the code to evaluate a scan to determine the FoV.

aburrell committed:
ENH: added location routines

Added routines that deal with geolocation and propagation paths.

aburrell committed:
ENH: improved the update_beam routine

Improved the update_beam routine `UpdateBeamFit` to include all important aspects of the beam initialization and beam-only groundscatter re-evaluation.

aburrell committed:
ENH: added more elements to scan code

Added more elements to the update_scan portion of the algorithm.

aburrell committed:
ENH: added routines to determine alt groups

Added routines to determine the appropriate virtual height limits of a selected group of data.  Not yet complete.

aburrell committed:
ENH: added sort utilities

Added sort utilities to the directory.

aburrell committed:
ENH: outlined least-squares fitting

Outlined the least-squares fitting portion of the algorithm.

aburrell committed:
ENH: update to altitude selection

Updated scan eval process by adding outline of the virtual height bin selection process.

aburrell committed:
ENH: added stat utilities

Added linear regression and other statistical utilities.

aburrell committed:
ENH: extended beam evaluation

Extended the azimuthal elevation angle evaluation.

aburrell committed:
ENH: finished update_scan routine

Finished initial write-up of update_scan.c.

aburrell committed:
BUG: fixed scan/beam assigment

Fixed the scan and beam assignment, updating parameters and assigning them from the correct structures.

aburrell committed:
ENH: added source files to makefile

Updated source file list and output name.

aburrell committed:
ENH: copied load_fit to local directory

Temporary (possibly) housing of load_fit in this directory.

aburrell committed:
ENH: created frequency band routine

Created a routine to determine frequency bands based on the loaded data and a desired band width.

aburrell committed:
ENH: updated main

Updated main to use newly enhanced routines, identifying next areas for development.

aburrell committed:
ENH: add help string

Added a help string to the help header.

aburrell committed:
ENH: added external files to local directory

Temporarily added external files that are needed locally as well.

aburrell committed:
BUG: added missing routine

Added a missing error specification routine.

aburrell committed:
BUG: fix bugs identified in compilation

Fixed bugs found when attempting to compile the new routine.

aburrell committed:
ENH: created routine for frequency exploration

Created routine to explore the transmission frequencies used, with the goal of establishing frequency band limits for the FoV determination routine.

aburrell committed:
BUG: fixed utility compilation bugs

Fixed bugs in `utils` revealed by compilation and created a header file to simplify routine declarations.

aburrell committed:
STY: removed commented code

Removed unused code lines that were commented out.

aburrell committed:
BUG: added mission option to help

Added a missing option to the help string.

aburrell committed:
BUG: fixed compilation errors in sort routines

Fixed compilation errors present in the sort routines and added a header file.

aburrell committed:
ENH: finished writing `select_alt_groups`

Finished writing `select_alt_groups` routine.

aburrell committed:
BUG: fixed some compilation errors

Fixed many of the compilation errors in `update_beam`.

aburrell committed:
BUG: fixed many compilation errors

Fixed many compilation errors in `update_scan`.

aburrell committed:
ENH: created multscan

Created a new multi-scan structure, since the RadarScan structure doesn't hold all of the necessary data.

aburrell committed:
ENH: updated BSID structures

Partial update of the BSID structures.

aburrell committed:
MAINT: removed `get_mult_scan_tfreq_band`

Removed `get_mult_scan_tfreq_band`, making it a part of its own routine.

aburrell committed:
ENH: added `slant_range` routine

Added a routine to calculate slant distance to the location utility.

aburrell committed:
MAINT: moved old routine

Temporarily moved an old routine to a new holding directory `old`.

aburrell committed:
ENH: saved testing files

Saved useful test files to the `old` directory.

aburrell committed:
BUG: fix bug in frequency cycling

Improved frequency band cycling, as previously done in frequency band branch.

aburrell committed:
ENH: temporarily added frequency band routine

Temporarily added frequency band determination routine to make_fov.

aburrell committed:
ENH: updated multBSID structure

Updated MultBSID structures and added Make/Free routines.

aburrell committed:
ENH: temporarily added cnvtcoord

Temporarily added cnvtcoord to avoid duplicating `slant_range` routine.

aburrell committed:
ENH: change elv routine inputs

Change the elevation angle routine inputs to use current structures.

aburrell committed:
ENH: updated basic structure

Updated the basic structure to include only the necessary information throughout the various routines.

aburrell committed:
ENH: temporarily remove makefile

Temporarily remove makefile from test utility routine that doesn't currently compile.

aburrell committed:
ENH: moved scan cycling

Moved the scan cycling to load and update one scan at a time, reducing memory cost.

aburrell committed:
BUG: fixed error in abs call

Used correct implementation of `fabs` for non-integer input.

aburrell committed:
ENH: added `gammaq` definition to header

Added the definition of `gamma` to header to allow use elsewhere.

aburrell committed:
BUG: fixed compilation errors

Fixed errors and warnings encountered during compilation.

aburrell committed:
ENH: added phase lag error

Added the phase lag error to the RadarCell structure, allowing elevation angle error calculations.

aburrell committed:
ENH: load elevation data

Load the elevation and phase lag data into the RadarCell structure.

aburrell committed:
MAINT: moved old files

Moved old files to a temporary directory.

aburrell committed:
MAINT: consolidated grid utilities

Put all utilities taken from `make_grid` into a single file.

aburrell committed:
BUG: fixed compilation errors

Fixed compilation errors and warnings.

aburrell committed:
ENH: added load_radar_site routine

Added a routine to load the RadarSite structure.

aburrell committed:
BUG: fixed frequency filtering

Fixed frequency filtering logic and moved the assignment of some of the scan pointers.

aburrell committed:
ENH: removed cfit support

CFit files don't have elevation angles, so don't need to be supported in this routine.

aburrell committed:
DOC: created make_fov xml doc file

Created the xml file needed for routine documentation, as well as help and error header files.

aburrell committed:
STY: moved files from `src`

Moved files from `src`, following RST conventions.

aburrell committed:
BUG: fixed compiler warnings

Fixed compiler warnings caused by typos.

aburrell committed:
BUG: fix PHI0 assignment

Only attempt to assign phi0 of the xrng data is available.

aburrell committed:
DOC: updated docs

Updated the make_fov doc, adding missing descriptions.

aburrell committed:
ENH: combined more grid utilities

Combined more of the grid utilities into a single file.

aburrell committed:
BUG: updated character assignment

Updated the character assignment for structure values.

aburrell committed:
BUG: updated vheight calc to double

Changed precision of virtual height calculation to double, as float caused notable errors.

aburrell committed:
BUG: fixed command line option bugs

Fixed various bugs in the command line options.  Opt expects character strings to be double pointers.

aburrell committed:
BUG: fixed load_radar_site bug

load_radar_site correctly assigns the site structure when it follows the same general structure as `RadarYMDHMSGetSite`.

aburrell committed:
TST: moved test makefile

Moved testing makefile to have a useful name.

aburrell committed:
TST: updated test file

Updated test file to test the relative maximum routine.

aburrell committed:
BUG: fixed alt selection bugs

Fixed bugs in the select altitude group routines including:
- initializing pointers using the wrong counter,
- adding missing variable definitions,
- narrowing acceptable MPFIT status definition, and
- fixing which variables are used in certain evaluations.

aburrell committed:
BUG: fixed scan cycle bugs

Fixed scan cycling bugs and added more error catches, including:
- test that CPID and channel are consistent across scans,
- fixed some beam number assignments,
- fixed bad iterator calls,
- moved the scan cycling call to the correct place,
- added a scan cycling closure call,
- added a check for correctly defined beam parameters, and
- removed debugging statements.

aburrell committed:
ENH: adapted clip method in `int_argrelmax`

Adapted clip method in `int_argrelmax` to allow maxima at the edges to be identified.

aburrell committed:
BUG: fixed load_fit_update_fov bugs

fixed bugs in `load_fit_update_fov`, including:
- initializing time variables,
- setting time variables,
- setting the RadarSite structure, and
- removing debug statements.

aburrell committed:
BUG: fixed bugs in `update_beam`

Fixed bugs in beam updating routine including:
- casting the earth radius in km,
- removing data where elevation angles can't be calculated, and
- fixing broken calls to front FoV elevation calculations.

aburrell committed:
BUG: added missing comma

Added a missing comma for an error message print statement.

aburrell committed:
ENH: added ascii write routine

Added a write routine with ascii output for easy debugging.

aburrell committed:
BUG: fixed referencing errors

Fixed pointer referencing errors and added range flag initialization.

aburrell committed:
ENH: added opposite FoV data as output

Added the opposing FoV data as testing output.

aburrell committed:
BUG: fixed location routines

Fixed location routines by:
- removing broken alias functionality from elevation angle routine,
- removing alias logic from the propagation path adjustment,
- adding hop logic to region setting function,
- moved slant range calculation to location file, since it differs from grid implementation.

aburrell committed:
BUG: removed alias logic

Removed the alias logic from the beam updating routine.

aburrell committed:
BUG: fixed memory leaks

Fixed memory leaks and potential errors identified by valgrind.

aburrell committed:
BUG: freed structure at the end

Freed the structure at the end of the main routine.  Also updated strict flag and F-region maximum height limits.

aburrell committed:
ENH: updated F_hmax

Updated `F_hmax` from 900 km to 750 km based on new evaluation of virtual height limits.

aburrell committed:
BUG: fixed bugs in scan evaluation

Fixed various bugs in the scan evaluation implementation.

aburrell committed:
ENH: added specific zeroing

Added zero-specification for values that may potentially be accessed elsewhere in the algorithm.

aburrell committed:
DOC: removed test code

Removed test code from the algorithm.

aburrell committed:
ENH: added UT consistency test

Added the single-beam consistency test as a function of range gate and UT.

aburrell committed:
TST: added debugging statement

Added a debug statement to help figure out why a clear 1.5F structure isn't being detected.

aburrell committed:
ENH: reduced duplicate code

Reduced duplicate code by creating a utility routine and added temporary debug statements.

aburrell committed:
TST: removed test line

Removed a debugging line.

aburrell committed:
BUG: relaxed slope requirement

Relaxed the slope requirement to be negative (flat is also good) and added an evaluation for swapping the FoV for a point when looking at inliers.

aburrell committed:
BUG: updated inlier evaluation

Fixed the behaviour of the inlier evaluation.

aburrell committed:
BUG: ensure `bad_fov` is set

Ensure `bad_fov` variable is set in all cases.

aburrell committed:
BUG: removed unused header

Removed an unused header.

aburrell committed:
TST: try adding a threading library

UNIX compilation errors related to this library identified.  Seeing if this works.

aburrell committed:
TST: added another header

Tried adding the "llist.h" header.

aburrell committed:
TST: added llist.h header

Added the llist.h header to the linear regression routine.

aburrell committed:
TST: added more include headers

Added more include headers.

aburrell committed:
MAINT: removed find_frequency routine

Removed the `find_frequency` routine from make_fov branch for a cleaner pull request.

aburrell committed:
MAINT: removed `read_fit` routine

Removed the `read_fit` routine from the make_fov branch for a cleaner pull request.

aburrell committed:
MAINT: removed old and test functions

Removed the old and test functions from the `make_fov` directory for a clean pull request.

aburrell committed:
MAINT: renamed utils

Renamed utils to be more descriptive.

aburrell committed:
MAINT: separated scan utilities

Separated the supporting routines from update_scan into their own file with a header.

aburrell committed:
DOC: updated file header

Updated the name and author for this header file.

aburrell committed:
TST: removed test line

Removed an old test line from the file.

aburrell committed:
MAINT: updated utility calls

Updated the various stat and scan utility calls.

aburrell committed:
MAINT: removed extra lines

Removed extra lines to make this file agree with develop.

aburrell committed:
DOC: updated the license

Updated the license to agree with develop.

aburrell commented:
I am having issues compiling this on CentOS release 6.10.  

Less worrying:  I am getting this warning that may or may not need to be addressed:
```
$RSTPATH/include/base/rtypes.h:49: warning: ISO C90 does not support ‘long long’
$RSTPATH/include/base/rtypes.h:50: warning: ISO C90 does not support ‘long long’
```

The error (shortened for legibility):
```
$RSTPATH/lib/libfitacf.3.0.a(llist.o): In function `llist_sort':
llist.c:(.text+0x2d2): undefined reference to `pthread_rwlock_unlock'
llist.c:(.text+0x2e6): undefined reference to `pthread_rwlock_wrlock'
$RSTPATH/lib/libfitacf.3.0.a(llist.o): In function `llist_reverse':
llist.c:(.text+0x355): undefined reference to `pthread_rwlock_wrlock'
llist.c:(.text+0x365): undefined reference to `pthread_rwlock_unlock'
$RSTPATH/lib/libfitacf.3.0.a(llist.o): In function `llist_concat':
... et cetera ...
$RSTPATH/lib/libfitacf.3.0.a(llist.o): In function `llist_create':
llist.c:(.text+0xe26): undefined reference to `pthread_rwlockattr_setpshared'
llist.c:(.text+0xe36): undefined reference to `pthread_rwlock_init'
llist.c:(.text+0xe42): undefined reference to `pthread_rwlockattr_destroy'
collect2: ld returned 1 exit status
make: *** [make_fov] Error 1
```

aburrell committed:
STY: removed extra whitespace

Removed extra whitespace from file.

aburrell committed:
ENH: only write with data

Added a logic catch to only write an output file if there's data.

aburrell committed:
TST: added error catches

Added error catches for:
- hardware file conflicting with local definitions,
- scan having too many beams,
- scan having duplicate beams.

aburrell committed:
BUG: added point limits

Added better point limits for different evaluation sections.

aburrell committed:
BUG: moved pointer init

Moved the `scan_new` pointer initialization inside the else loop where it is needed, instead of at the top of the routine where it may not be needed.

aburrell committed:
STY: updated inline comments

Updated the inline comment style.

aburrell committed:
DOC: updated comments

Added and improved comments.

aburrell committed:
STY: improved spacing

Made spacing more consistent and added a free call for the radar parameter structure.

aburrell committed:
BUG: removed unused input variable

Removed an unused input variable.

aburrell committed:
BUG: fixed memory leaks

Fixed unexpected memory links by using arrays instead of pointers for grouped scan data.

aburrell committed:
ENH: added argsort for int

Added an argument sort function for integer input.

aburrell committed:
ENH: added multi-peak gaussian

Added routines to fit multi-peaked Gaussian functions.

aburrell committed:
ENH: updated fit criteria

Updated the criteria for selecting a good linear fit to the elevation angles.

aburrell committed:
ENH: updated alt group selection

Updated the altitude group selection to use a mult-peak Gaussian fitting process and configure the least-squares fitting boundaries.

aburrell committed:
ENH: added mult-beam time option

Added an option to consider multiple beams when examining the temporal variations.

aburrell committed:
ENH: updated azimuth eval

Updated the azimuth eval to consider the standard deviation, but allow equality.

aburrell committed:
ENH: changed swap criteria

Made the criteria for swapping fields of view stricter.

aburrell commented on codebase/superdarn/src.bin/tk/tool/make_fov.1.0/eval_groundscatter.c:
```diff
@@ -0,0 +1,190 @@
+
+  return(frac);
+}
+			    
+	  
+
+  
```
```suggestion

```

aburrell commented on codebase/superdarn/src.lib/tk/scan.1.7/include/scandata.h:
```diff
@@ -22,6 +22,8 @@ You should have received a copy of the GNU General Public License
 
 Modifications:
+
+  AGB Jul 2021 - Added phase lag error to the RadarCell structure
```
```suggestion
 Jul 2021 Angeline G. Burrell: Added phase lag error to the RadarCell structure
```

aburrell committed:
MAINT: fixed small style issues

Fixed extra lines at the end of a file and the modification style on a header.

aburrell committed:
DEL: removed temporary files

Removed temporary files that contain routines now available in libraries.

aburrell committed:
DOC: updated licenses

Updated licenses to specifically reference RST.

aburrell committed:
MAINT: removed old declaration

Removed a function declaration that is no longer needed now that it is available through one of the libraries already loaded in the header files.

aburrell committed:
MAINT: updated use of libaries

Updated the main routine and makefile to use functions now available through libraries.

aburrell committed:
STY: moved elv calc to library

Moved the elevation angle calculation with lobe taken into account into the elevation angle library.  Added the elevation angle header to the appropriate files and library flag to the makefile.

aburrell committed:
STY: moved position routines

Moved the virtual height and slant range routines from the local location.c file to the rpos library.  Updated the license description in the rpos library makefile.

aburrell committed:
BUG: added missing libraries

Added missing libraries needed by the new elevation angle routine.

aburrell committed:
MAINT: removed excess whitespace

Removed excess whitespace from the scan library makefile.

aburrell committed:
ENH: created a propagation library

Created a library to hold routines related to propagation and region calculations and evaluations.

aburrell committed:
ENH: utilized propagation library

Updated the make_fov code to utilize the new propagation library.

aburrell committed:
DOC: updated eval GS docs

Updated the license and function descriptions in `eval_groundscatter.c`.

aburrell committed:
ENH: added sort library

Added a library with sorting functions.  Updated make_fov to use the new sorting library.

aburrell committed:
ENH: added fitmultbsid library

Added a library for the FitMultBSID structures and supporting functions.  Updated make_fov to use this new library.

aburrell committed:
ENH: added stats library

Added a statistics library to the analysis code base.  Updated make_fov to use the statistics library.

aburrell committed:
ENH: moved gamma functions to stats

Moved the gamma functions from the fitacf 3.0 library to the stats library.

aburrell committed:
DEL: removed temporary file

Remove a file kept temporarily for development.

aburrell committed:
ENH: adapted to use new gamma

Adapted the make_fov routines to use the gamma functions in the stats library.

aburrell committed:
ENH: added linear fit library

Added a library for linear fit routines.  Updated make_fov to use this library.

aburrell committed:
BUG: fixed include statement

Changed from "import" to "include", the correct C-library statement!

aburrell committed:
ENH: added `load_radar_site` routine

Added the `load_radar_site` routine to the radar library.  Updated make_fov and make_grid.2.0 routines to use this new function.

aburrell committed:
BUG: removed duplicate line in makefile

Removed duplicate line in makefile and updated documentation in other files.

aburrell commented on codebase/superdarn/src.bin/tk/tool/make_grid.2.0/make_grid.c:
```diff
@@ -26,6 +26,7 @@ along with this program. If not, see <https://www.gnu.org/licenses/>.
 Modifications:
  2021-01-11 Marina Schmidt fixing concatenation bug with start/end time, and deprecated -c flag
  2021-08-27 Angeline G Burrell removed duplicate code now present in libraries
+ 2021-10-04 Angeline G Burrell removed duplicate code now present in radar librarie
```
```suggestion
 2021-10-04 Angeline G Burrell removed duplicate code now present in radar library
```

aburrell committed:
MAINT: fixed spelling

Fixed spelling of 'library'.

aburrell committed:
ENH: added binary write routine

Added a placeholder for the `FitMultBSID` binary write routine.  "HERE" in comments denotes place to pick up work.

aburrell committed:
ENH: finished binary write routine

Finished writing the binary writing routine for the FitMultBSID data and added it to the library.

aburrell committed:
ENH: updated input options

Updated make_fov input options by:
- making all davitpy input options also RST command line options,
- updating the documentation, and
- adding an option for binary or ASCII output.

aburrell committed:
DEL: removed savefile

Removed an emacs savefile.

aburrell reviewed (commented):
Also need to remove "#rawrite.c#" file.

aburrell committed:
STY: added missing whitespace

Added missing whitespace to maintain style consistency.

aburrell committed:
DOC: updated make_fov examples

Updated the examples in the make_fov documentation.

aburrell committed:
STY: whitespace

Added and removed whitespace to improve legibility.

aburrell committed:
STY: clarified NULL casting

Improved NULL casting to reduce valgrind potential memory errors.

aburrell committed:
ENH: added `RadarParmReset` function

Added the RadarParmReset function to reduce potential instances of duplicate code.

aburrell committed:
STY: reduced length of strings

Reduced the length of strings, to not carry around more information than is strictly necessary.

aburrell committed:
MAINT: updated default group

Updated the default group ID to be 'U' for unset groups.

aburrell committed:
STY: standardized whitespace

Standardized the spacing for this routine.

aburrell committed:
BUG: added free for variable

Added a missing `free` for a malloc'ed variable.

aburrell committed:
BUG: fixed bugs in binary write

Fixed bugs causing segfaults in the binary writing routine.

aburrell committed:
BUG: fixed write bugs

Fixed bugs in the binary write routine:
- fixed name of the spectral width, and
- added a beam number variable for each scan.

aburrell committed:
ENH: added a binary read routine

Added a binary read routine for the BSID files.

aburrell committed:
ENH: added FitMultBSID read tool

Added a tool to convert FitMultBSID files from binary to ASCII.

aburrell committed:
BUG: added debug statements

Fixed some scan initialization bugs, removed an unused variable, and added debugging statements.

aburrell committed:
BUG: debug attempts

Debugging attempts to make sure binary can be read back into data structure.

aburrell committed:
BUG: added missing libraries

Added missing libraries to the fitmultbsid header and the reformatting tool.

aburrell committed:
BUG: added stats library

Added the stats library to the fitacf makefile.

aburrell committed:
BUG: added sort library

Added the sort library to the fitacf makefile.

aburrell committed:
BUG: fixed include

Fixed the include scope for make_fit.

aburrell committed:
ENH: added sort library

Explicitly add the sort library to the leastsquares.c file in FitACF 3.0.

aburrell committed:
BUG: changed library order

Changed the order of the libraries.

aburrell committed:
BUG: fixed binary reading

Fixed the MultBSID binary file reading routine.

aburrell committed:
BUG: fixed string initialization

Added string initialization to unset values before reading in data.

aburrell committed:
BUG: updated time setting

Updated the time setting to avoid duplicate values and ensure time reading is simple.

aburrell committed:
STY: removed extra whitespace

Removed the extra whitespace from a header file.

aburrell committed:
DOC: added FitMultBSID documentation

Added development gotchas and a summary of the FitMultBSID structures to the  docs.

aburrell commented on codebase/superdarn/src.bin/tk/tool/make_fov.1.0/test_ut_fov_struct.c:
```diff
@@ -0,0 +1,609 @@
+ * @params[in] vb         - Verbosity flag
+ *             vbuf       - Verbosity buffer
+ *             nbms       - Number of beams used in each time eval (1)
+ *             min_frac   - Minimum fraction of possible backscatter points
+                            needed in the RG/UT box to test the FoV (0.1)
+ *             ut_box_sec - UT box size in decimal seconds (1200.0)
```
```suggestion
 *             min_frac   - Minimum fraction of possible backscatter points
 *                          needed in the RG/UT box to test the FoV (0.1)
 *             ut_box_sec - UT box size in decimal seconds (1200.0)
```

aburrell committed:
DOC: update docstring

Added missing asterisk to the docstring.

aburrell commented:
@egthomas @ecbland this is now ready for review!

aburrell committed:
BUG: fixed multi-file reading

Fixed bugs involved in loading data from multiple binary files including:
- reading in multiple filenames from the command-line input,
- reading in the binary header data, and
- cycling through the new scans.

pasha-ponomarenko commented:
@aburrell, I am a bit confused about how to test this routine. 
I have installed the branch and ran _make_fov_ on a FITACF file. The output seems to be in DMAP format rather than ASCII mentioned at the beginning of the issue. I can also see that it is a FITACF-like format but with extra fields. 
Could you please provide a bit more comprehensive instructions for a non-initiated reviewer like me? :-)

aburrell commented:
@pasha-ponomarenko before I give you detailed instructions (which I am happy to do) could you try following the examples I wrote up in the local routine docs?  I think it would be good to have someone unfamiliar with the routine, but a competent RST user try to follow them so I can use your feedback to make them clearer.

But to answer a quick question you raised, there are two output formats: ascii and a dmap binary.  It is basically fitacf + some fields.  Not all fields are used by this routine, some are there in preparation for the rest of the backscatter ID processing that I will code up following this pull.

mts299 commented:
@aburrell sorry what are the new fields called? 

Make sure to add them to pyDARNio so people can read them :)

pasha-ponomarenko commented:
@aburrell, do you mean this one?
https://github.com/SuperDARN/rst/blob/make_fov/codebase/superdarn/src.bin/tk/tool/make_fov.1.0/docs/make_fov.doc.xml

When I try to open it in a browser, I get this error:
![image](https://user-images.githubusercontent.com/5315016/152373134-0a5784a3-9763-473f-88af-1ccc7149fd12.png)

It might be something very simple, but ...

aburrell commented:
I do mean that one!  Unfortunately I do not know XML, so I made the docs by copy-pasta + add sauce.  Can someone with XML knowledge take a look at that file and let me know what might be wrong?  @pasha-ponomarenko do you know anyone with this skill set?

@mts299 the new fields are all in the new structure housed in `$RSTPATH/codebase/superdarn/src.lib/tk/fitmultbsid.1.0/`.    All the parameters are explained in the docs in that file, so if you don't mind trying to read them and coming back with questions, I'd appreciate your partial review!

mts299 commented:
I will do it next week (maybe), currently cannot stare at a screen due to a concussion. Was just looking for quick things to do or ask.

pasha-ponomarenko commented:
Apparently, there is a know issue with Github: one should never download a web file from Git as it will contain the Git webpage information as well: 
https://github.com/pm4py/pm4py-core/issues/247#issuecomment-907742973 

However, when I tried to open the same file from an RST package installed on my Linux desktop, it just showed a "raw" ASCII content and the following error(?) message: 
"This XML file does not appear to have any style information associated with it. The document tree is shown below. "
This happens on both Linux and Windows.

@aburrell, are you able to see the document properly in your web browser?

aburrell commented:
So, opening any of the files in a web browser doesn't do anything but display the code.  I guess the good news is that my xml behaves the same as the rest of the xml.  None of them open.

Given that it seems none of us have the knowledge to resolve this issue, here's what you need to run the code:

Create ascii output: make_fov -ascii <filename>
Use your own tdiff: make_fov -update-tdiff -tdiff <value_in_microseconds> <filename>
Specify your own frequency band: make_fov -tfmin <value in kHz> -tfmax <value in kHz> <filename>
Specify channel a: make_fov -cn a <filename>

Because you need to look at the range and azimuthal variations, you need to separate frequency bands for the analysis.  For best results, use the TDIFF suited to the choses radar channel and frequency! 

To convert the dmap binary to ascii, use `fitmultbsidtoascii`

pasha-ponomarenko commented:
Yep, the XML business looks like a separate issue. 
:-(
Thanks for the clarification. Do you have a description of the extra fields in a more robust format like DOC or PDF? Or simple ASCII?

aburrell commented:
No, I put all the documentation in xml since that was the RST standard.

The most important thing is FOVFLG, which is 1 for the front FoV, -1 for the rear FoV, and 0 if a FoV could not be cleanly identified.  GFLG is also updated so that it is -1 if it was formerly groundscatter and doesn't look like groundscatter.

ecbland commented:
@aburrell Now I am confused. Here's a screenshot from following the instructions I gave in my previous comment. Is this what @pasha-ponomarenko was looking for?

![image](https://user-images.githubusercontent.com/22493623/152403716-27c0c573-ab7d-4054-98a6-5a8053bc079e.png)

aburrell commented:
> Is this what @pasha-ponomarenko was looking for?

Yes, @ecbland that is it!

aburrell committed:
STY: removed sprintf string appending

aburrell committed:
BUG: fixed mult-file read

pasha-ponomarenko commented:
@aburrell, I might miss something, but there seem to be fewer fields in a binary output of make_fov than in its ASCII one, e.g., MED_* and GRPFLG, GRPNUM, GRPID are all missing.

pasha-ponomarenko commented:
OK, here are some preliminaries.
I used 20150223.2001.00.sas.fitacf (v3, twofrequency mode) which contains both ionospheric and ground scatter. I also selected SAS data because the ground scatter here is normally dominated by signals received through the grating lobe. Here are plots for the 10 MHz band of the twofrequency mode. Importantly, for this interval there was no detectable tdiff offset for this frequency band. The data show both ionospheric and ground scatter, and ground scatter coming from rather high elevation angles which is indicative of the "back" lobe scatter.

![image](https://user-images.githubusercontent.com/5315016/153469180-01c4a424-9662-44f5-9d50-0e487a9e7bb4.png)

![image](https://user-images.githubusercontent.com/5315016/153469319-43cb4727-5373-464d-acc7-18ae8e6eed77.png)

I processed this data for frequencies between 10 and 11 MHZ using following command line options:
`make_fov -vb -tfmin 10000 -tfmax 11000 20150223.2001.00.sas.fitacf > 20150223.2001.00.10.sas.fov`
Then I converted DMAP *.fov file into NetCDF with following steps:
- Analysed DMAP structure and produced supplementary files: 
`dmaptocdl 20150223.2001.00.10.sas.fov fov.cdl fov.cdlmap `
- Adjusted the field lengths in the "mask" file fov.cdl to 75.
- Generated an empty NetCDF file
`ncgen fov.cdl -o fov_10.netcdf`
- Filled the empty file with data from DMAP file:
`dmaptoncdf 20150223.2001.00.10.sas.fov fov.cdlmap fov_10.netcdf`

After that used IDL to read and plot NetCDF FOV data:

Velocity
![image](https://user-images.githubusercontent.com/5315016/153470818-7f91deb2-5380-4bfc-b78e-ee742f6c6463.png)

Elevation
![image](https://user-images.githubusercontent.com/5315016/153470898-f298f0c0-5534-4875-9104-3a889fcfe3d0.png)
FOV
![image](https://user-images.githubusercontent.com/5315016/153470950-19434905-26d1-418a-b204-2b67296c23ce.png)
GSCT
![image](https://user-images.githubusercontent.com/5315016/153470995-a8f0c2c2-379f-44ce-a3b9-39654dccb4f2.png)
SCT
![image](https://user-images.githubusercontent.com/5315016/153471069-5a7ce9bf-ded0-48f1-b4b3-221a0e4d1621.png)

pasha-ponomarenko commented:
My understanding is that the majority of the 1.5 ionospheric scatter above rage gate 40 is removed due to one of the selection criteria (exceeding the maximum virtual height?). 
The FOV determination seems to be a bit shaky for the majority of the ground scatter before 21:30 UT. I guess it might be related to an apparent partial overlap between the front- and "back"-lobe ground scatter. Indeed, the closer edge of the GS band between gates 18 and 25 shows very high elevation angle values in excess of 35 deg followed by a sharp boundary at  with elevation angle values dropping to  20-25 deg at gates ~25-32. 

@aburrell, please correct me if I am wrong here.
BTW, I am not sure what SCT flag means.

aburrell committed:
BUG: fixes for new tdiff struct

Made changes needed to use the new channel-dependent tdiff.

aburrell committed:
DOC: update elevation docstring

Update the parameter description for the lobe dependent elevation calculation.

aburrell committed:
DOC: added prop warning

Added a propagation path warning for the users.

aburrell commented:
Ok, I played around with things and the main reason some things aren't working well is the propagation grouping.  I applied the HAN defaults (basically the same, but `-fhmax 900`) and got much more of the ionospheric scatter back.  Fine tuning the propagation virtual height ranges would further improve the results.  I added a note to the docs to make sure users know how important this is, and know where the defaults come from.

Looking at the azimuthal variations, it makes sense that much of the groundscatter returns from the front FoV, since the variations are consistent across the beams. 

![sas_bm15_fhmax900_test](https://user-images.githubusercontent.com/7045886/162502034-fc478c5e-259d-4902-a589-3182f3cf7cb1.png)

aburrell commented:
> BTW, I am not sure what SCT flag means.

It's the `sct` parameter from the fitacf structure, I believe it just flags whether the scatter is valid or not.

pasha-ponomarenko commented:
> It's the sct parameter from the fitacf structure, I believe it just flags whether the scatter is valid or not.

I don think there is such a parameter in the fitacf structure 
https://radar-software-toolkit-rst.readthedocs.io/en/latest/references/general/fitacf/, 
that is why I asked 
:-P
It does not look like a quality flag...

pasha-ponomarenko commented:
@aburrell , What language did you make your plots in? Python? IDL?
Is there a way to integrate some plotting into this package (or utilise some existing RST plotting routines) in order to make tester's/user's life a bit easier?
:-{

aburrell commented:
> I don think there is such a parameter in the fitacf structure

There is!  Just grep "sct" in $RSTPATH/include/superdarn/*.h and you'll see which structures it's in.  You can also examine the code to see how it's used.  It may not be included in the user-description of the output, but it's a code structure parameter.

> What language did you make your plots in

Python.  I don't know how to plot in C, so I won't be adding this to RST.  They're just RTI plots like the ones you made, but further separated by hop or propagation path.  The plots you made were fine, they confirmed the code is working as it should be!

aburrell committed:
ENH: added idl.tar.gz to .gitignore

Added the IDL tarball created when computers without IDL compile RST to the list of ignored files.

aburrell committed:
BUG: init pointer

Added pointer initialization.

aburrell committed:
ENH: added extreme value analysis

Added an evaluation for all elevation angles having extreme values (at the upper edge of elevation angle measurements).

aburrell committed:
ENH: added extreme fov support

Added support for extreme fov evaluation.

aburrell committed:
ENH: updated extreme fov and fov order

Updated the order in which the two fields of view are evaluated (front first) and incorporated support for extreme elevation angle evaluation.

aburrell committed:
ENH: added new gflg

Added a new gflg to the updated data structure.  Preserves original groundscatter flag `gsct`.

aburrell committed:
ENH: initialize new gflg value

Initialize a new groundscatter flag value with the original groundscatter flag value when loading scans.

aburrell committed:
ENH: update beam evaluation

Update beam evaluation to use new groundscatter flag instead of the original groundscatter flag.

aburrell committed:
ENH: updated groundscatter evaluation

Updated the groundscatter flag evaluation to alter the new flag value and maintain the old one for continuity with previous code and ease of comparison.

aburrell committed:
BUG: re-add make_fit libraries

Some libraries that `make_fit` needs were removed during the process of resolving merge conflicts.

aburrell committed:
BUG: fixed typo in structure

Fixed a typo in the structure reference.

aburrell committed:
BUG: removed unused variables

Removed unused variables from the `make_fov` functions.

aburrell committed:
BUG: updated library version

Updated the mpfit library version.

aburrell commented on codebase/superdarn/src.lib/tk/fitmultbsid.1.0/include/fitmultbsid.h:
```diff
@@ -0,0 +1,154 @@
+
+struct CellBSIDFlgs
+{
+  int gflg;
```
New groundscatter flag.

aburrell commented on codebase/superdarn/src.lib/tk/fitmultbsid.1.0/src/read_binary.c:
```diff
@@ -0,0 +1,582 @@
+    {
+      aptr = ptr->arr[i];
+
+      if((strcmp(aptr->name, "gsct") == 0) && (aptr->type == DATASHORT)
+	 && (aptr->dim == 1))
+	{
+	  for(j = 0; j < aptr->rng[0]; j++)
+	    bm->rng[slist[j]].gsct = aptr->data.sptr[j];
+	}
+      else if((strcmp(aptr->name, "gflg") == 0) && (aptr->type == DATASHORT)
+	 && (aptr->dim == 1))
+	{
+	  for(j = 0; j < aptr->rng[0]; j++)
+	    bm->rng_flgs[slist[j]].gflg = aptr->data.sptr[j];
+	}
```
Now with two groundscatter flags (original and new).

aburrell commented on codebase/superdarn/src.lib/tk/fitmultbsid.1.0/src/write_ascii.c:
```diff
@@ -0,0 +1,151 @@
+    {
+      sprintf(scan_info, "#STID");
+      sprintf(bm_info, "DATE TIME INTT_US BMNUM BMAZM CPID NAVE FRANG RSEP RXRISE FREQ NOISE ATTEN CHANNEL NRANG");
+      sprintf(rng_info, "RG GFLG NEW_GFLG FOVFLG FOV_PAST GRPFLG GRPNUM GRPID P_0 P_0_ERR V V_ERR W_L W_L_ERR P_L P_L_ERR PHI0 PHI0_ERR ELV ELV_LOW ELV_HIGH VH VH_ERR VH_METHOD REGION HOP DIST MED_P_0 MED_P_0_ERR MED_V MED_V_ERR MED_W_L MED_W_L_ERR MED_P_L MED_P_L_ERR MED_PHI0 MED_PHI0_ERR OPP_ELV OPP_ELV_LOW OPP_ELV_HIGH OPP_VH OPP_VH_ERR OPP_VH_METHOD OPP_REGION OPP_HOP OPP_DIST");
```
New groundscatter output is provided alongside original.

aburrell commented on codebase/superdarn/src.lib/tk/fitmultbsid.1.0/src/write_binary.c:
```diff
@@ -0,0 +1,380 @@
+
+  /* Initiallize the range gate information */
+  slist    = DataMapStoreArray(ptr, "slist", DATASHORT, 1, &ngood, NULL);
+  gsct     = DataMapStoreArray(ptr, "gsct", DATASHORT, 1, &ngood, NULL);
+  gflg     = DataMapStoreArray(ptr, "gflg", DATASHORT, 1, &ngood, NULL);
```
New groundscatter flag provided alongside original.  Original keeps the name, new flag has new name.

aburrell commented on tables/superdarn/hdw/hdw.dat.dce:
```diff
@@ -10,3 +10,106 @@
+#     through transmitter to phasing matrix input. If the signal from the
+#     interferometer comes first, then tdiff < 0. Units are in decimal
+#     microseconds)
+# 12) Phase sign (Cabling errors can lead to a 180 degree shift of the
+#     interferometry phase measurement. +1 indicates that the sign is
+#     correct, -1 indicates that it must be flipped.)
+# 13) Interferometer offset  (Displacement of midpoint of
+#     interferometer array from midpoint of main array. This is given in
+#     meters in Cartesian coordinates. X is along the line of antennas with
+#     +X toward higher antenna numbers, Y is along the array normal
+#     direction with +Y in the direction of the array normal. Z is the
+#     altitude difference, +Z up.)
+#
+# More analog receiver information
+#
+# 14) Analog Rx rise time (Time given in microseconds. Time delays of
+#     less than ~10 microseconds can be ignored. If narrow-band filters are
+#     used in analog receivers or front-ends, the time delays should be
+#     specified.)
+# 15) Analog attenuation stages (Number of stages. This is used for
+#     gain control of an analog receiver or front-end.)
+# 16) Maxinum of range gates used (Number of gates. This is used for
+#     allocation of array storage.)
+# 17) Maximum number of beams to be used at a particular radar site.
+#     (Number of beams. It is important to specify the true  maximum. This
+#     will assure that a given beam number always points in the same
+#     direction. A subset of these beams, e.g. 8-23, can be used for
+#     standard 16 beam operation.)
+#
+# **********************************************************************
+# ==Notes==
+#
+#
+# **********************************************************************
+#
+#  1   2      3        4         5       6      7      8    9   10   11   12  13(1)  13(2)  13(3)  14   15  16 17
+#
+# UNTIL 26/Oct/2999 11:00:01.0
+   96 2999 25786801  -75.090  123.350  3233.0  115.0 -3.24  1   10 -0.151  1    0.0  -90.0    0.0 100.0  7  75 16
+# EOF
```
```suggestion
```

aburrell commented on codebase/superdarn/src.bin/tk/tool/make_fov.1.0/update_scan.c:
```diff
@@ -0,0 +1,580 @@
+			  bm_new->rng[irg].elv    = bm_old.rng[irg].elv;
+
+			  /* Initialize the range gate flags */
+			  bm_new->rng_flgs[irg].gflg = bm_new->rng[irg].gsct;
```
New gflg is initialized here.

aburrell commented on codebase/superdarn/src.bin/tk/tool/make_fov.1.0/update_beam.c:
```diff
@@ -0,0 +1,215 @@
+    }
+
+  /* Update the groundscatter flag */
+  EvalGroundScatter(beam);
```
Updates to new gflag happen here.

aburrell commented on codebase/superdarn/src.bin/tk/tool/make_fov.1.0/eval_groundscatter.c:
```diff
@@ -0,0 +1,175 @@
+ *        - at least 3 groundscatter points are found in a 10 range gate box
+ **/
+
+void EvalGroundScatter(struct FitBSIDBeam *beam)
```
Updated this function to leave the original `gsct` flag alone.

aburrell reviewed (commented):
@egthomas the compilation is failing for two routines that I can't see any connection to:
- rpos_v2_test.1.0
- make_sim.1.0

In both cases I think it's because the rpos.1 library is not being made correctly.  Any idea why?  I tried a few things without success.

aburrell committed:
DOC: Update tables/superdarn/hdw/hdw.dat.dce

Remove lines added during bad merge resolve.

aburrell committed:
REV: remove odd file

Remove file that shouldn't be there?

aburrell committed:
Merge branch 'make_fov' of https://github.com/SuperDARN/rst into make_fov

aburrell committed:
REV: file should be there

Copied file back from develop.

aburrell committed:
STY: removed tabs

Removed tabs from the end of each line.

egthomas committed:
Merge branch 'develop' into make_fov

egthomas committed:
Fixing compilation errors in make_fov branch

ksterne commented:
Thanks @egthomas for updating this!  A quick bump to ask is this ready to be merged in now?

ksterne commented:
Hello all, giving this another bump.  Can this be merged in now?  Maybe this is a quick discussion during the next DAWG meeting, though it would be great to see movement on it before then.

egthomas commented:
Sorry @ksterne and @aburrell for the slow response.  There are several core libraries that received changes / refactoring as part of this pull request, so since it has been so long I think it would be prudent to make sure the associated top-level routines like `make_fit` and `make_grid` still give consistent results.  There may also be some overlap with existing elevation angle / geolocation routines where we might be able to consolidate code.  So I would suggest we make this the focus of our next meeting to discuss a concrete plan for a few final tests - does that sound reasonable?
